src/bencode/decode.rs

```diff
@@ -1,51 +1,80 @@
+use std::collections::BTreeMap;
 use bencode::{Bencode, BString, BInt, BList, BDict, DecodeError, DecodeErrorKind};
 
-pub fn bstring_decode(bytes: Vec<u8>) -> Result<(BString, Vec<u8>), DecodeError> {
-    let mut length_str = String::new();
-    let mut position = 0u64;
-    if bytes.len() > 0 && (bytes[0] as char).is_digit(10) {
-        'topl: for ch in &bytes {
-            position += 1;
-            if (*ch as char) == ':' {
-                for byte in &bytes {
-                    if (*byte as char) != ':' {
-                        length_str.push((*byte as char));
-                    } else {
-                        break 'topl;
-                    }
-                }
-            }
-        }
+pub fn belement_decode(bytes: &[u8], position: &mut usize) -> Result<Bencode, DecodeError> {
+    if bytes.len() == 0 || *position >= bytes.len() {
```

Comment:
The advantage of having `position` not be a reference is that constant dereferencing can be avoided.